File: contest/LeetCode_Weekly_Contest_202/Minimum_Number_of_Days_to_Eat_N_Oranges.py
class Solution(object):
    def minDays(self, n):
        """
        :type n: int
        :rtype: int
        """
        
        dp = {}
        
        def dfs(n):
            if n == 0: return 0
            elif n == 1: return 1
            elif n == 2: return 2
            
            if n in dp: return dp[n]
            
            ret = min(dfs(n / 3) + n % 3, dfs(n / 2) + n % 2) + 1
            dp[n] = ret
            
            return ret
        
        return dfs(n)

File: contest/LeetCode_Weekly_Contest_202/Minimum_Number_of_Days_to_Eat_N_Oranges.java
class Solution {
    
    HashMap<Integer, Integer> dp;
    
    int dfs(int n) {
        if (n == 0) return 0;
        else if (n == 1) return 1;
        else if (n == 2) return 2;
        
        if (dp.containsKey(n)) return dp.get(n);
        
        int ret = Math.min(dfs(n/3) + n % 3, dfs(n/2) + n % 2) + 1;
        
        dp.put(n, ret);
        return ret;
    }
        
    public int minDays(int n) {
        dp = new HashMap<Integer, Integer>();
        return dfs(n);
    }
}
